### PR TITLE
Adding test cases to resolve Issues 250 and 251

### LINKED ISSUE #250: Create a duplex scenario test to pass a complex type in callback not used in service contract
We need a test where the duplex "Ping" callback passes a complex type that is not used in any service contract.  This will verify whether a serializer has been generated for it in NET Native.  It is critical this type not appear in any service contract's operations, or the serializer will have been generated from that usage.  It is also critical that no test explicitly serializes or deserializes this type, or the toolchain will create a serializer based on that usage.

### LINKED ISSUE #251: Create a duplex scenario test to use Xml in the callback contract
First, determine whether it is a supported scenario for a duplex callback contract to specify an [XmlSerializerFormat] attribute or an XmlSerializerOperationBehavior so that its "Ping" callback passes the value as xml rather than data contract.  And for the type, use a complex type not used in any service contract.

If supported, create a scenario test that does this.  This is to verify 3 things:
1. An xml serializer will be generated in NET Native for this type
2. The ContractDescription creation correctly recognizes this operation behavior on a callback contract
3. The e2e serialization+deserialization works correctly

### PR DESCRIPTION
**Pull request by StephenBonikowsky**
* Fixes #250 and Fixes #251

### CODE REVIEW

**Comment by StephenBonikowsky** [comment]
These tests have not been run against NET Native yet.
The Xml test fails, see Issue #292

**Comment by StephenBonikowsky** [comment]
@dotnet-bot retest this please

File: src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs [hunk 1]
```diff
@@ -803,4 +831,26 @@ public void OnPingCallback(Guid guid)
...
     }
+
+    public void OnXmlPingCallback(XmlCompositeType xmlCompositeType)
+    {
+        // Set the result in an async task with a 100ms delay to prevent a race condition
+        // where the OnPingCallback hasn't sent the reply to the server before the channel is closed.
+        Task.Run(async () =>
+        {
+            await Task.Delay(100);
+            _xml_tcs.SetResult(xmlCompositeType);
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs]: Just a question -- why 100ms?  Don't block on this, I was just wondering if there is some other way to detect when it's okay to set the TCS.

File: src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs [hunk 2]
```diff
@@ -793,6 +797,30 @@ public Guid CallbackGuid
...
+            if (_xml_tcs.Task.Wait(ScenarioTestHelpers.TestTimeout))
+            {
+                return _xml_tcs.Task.Result;
+            }
+            throw new TimeoutException(string.Format("Not completed within the alloted time of {0}", ScenarioTestHelpers.TestTimeout));
+        }
+    }
+
+    public ComplexCompositeType DataContractCallbackGuid
+    {
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs]: Possibly add a comment that no service contract is permitted to use this type.  It is reserved for the scenario that is testing that its appearance in a callback contract generates a serializer.  Probably same for the one above.

File: src/System.Private.ServiceModel/tools/test/SelfHostWcfService/TestResources/DuplexResources.cs [hunk 3]
```diff
@@ -57,4 +57,38 @@ protected override Binding GetBinding()
...
+    }
+
+    internal class DuplexCallbackXmlComplexTypeResource : EndpointResource<WcfDuplexService, IWcfDuplexService_Xml>
+    {
+        protected override string Protocol { get { return BaseAddressResource.Tcp; } }
+
+        protected override int GetPort(ResourceRequestContext context)
+        {
+            return context.BridgeConfiguration.BridgeTcpPort + 5;
+        }
```
Comment (roncain) [inline src/System.Private.ServiceModel/tools/test/SelfHostWcfService/TestResources/DuplexResources.cs]: I see you found my new way of choosing TCP ports :smile: .  Not now, but let's consider making some enum somewhere that gives them names.

**Comment by roncain** [comment]
LGTM -- all my comments are optional before merging.

File: src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs (hunk 1 above)
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs]: Not sure, I think either Matt or Jason added this functionality to the existing 'OnPingCallback' which I then adopted., I can confirm the race condition does occur.
I'm not sure how else we could do it, I can't see a way to make the cleanup code smarter for this particular scenario, but there may be a solution there.

File: src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs (hunk 2 above)
Comment (StephenBonikowsky) [inline src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestTypes.cs]: Done, I also created unique versions as I was re-using these types. I agree that having unique/use once types is better/safer than trying to figure out how they are safe to use by other tests without messing up these tests.

**Comment by roncain** [comment]
Still LGTM @StephenBonikowsky and it passes all InnerLoop and OuterLoop tests.

**Pull request merged**